File: BACKEND/src/main/java/com/macro/macro/Service/ChamadosService.java
package com.macro.macro.Service;


import com.google.api.core.ApiFuture;
import com.google.cloud.firestore.*;
import com.google.firebase.cloud.FirestoreClient;
import com.macro.macro.Model.Chamado;
import com.macro.macro.Model.Microservico;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutionException;

@Service
public class ChamadosService {
    public static final String COL_NAME = "Chamados";

    public Chamado save(Chamado data) {
        try {
            Firestore db = FirestoreClient.getFirestore();

            DocumentReference docRef = db.collection(COL_NAME).document();
            data.setId(docRef.getId());

            data.setCreatedAt(String.valueOf(LocalDate.now()));

            docRef.set(data).get();

            return data;

        } catch (Exception e) {
            throw new RuntimeException("Erro ao salvar chamado");
        }
    }

    public Chamado update(Chamado data) {
        try {
            Firestore db = FirestoreClient.getFirestore();

            DocumentReference docRef =
                    db.collection(COL_NAME).document(data.getId());

            DocumentSnapshot snapshot = docRef.get().get();

            if (!snapshot.exists()) {
                throw new RuntimeException("Chamado não encontrado");
            }

            data.setUpdatedAt(String.valueOf(LocalDate.now()));

            docRef.set(data).get();

            return data;

        } catch (ExecutionException | InterruptedException e) {
            throw new RuntimeException(e);
        }
    }

    public Chamado findById(String id) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();

        DocumentSnapshot doc = db.collection(COL_NAME).document(id).get().get();

        if (!doc.exists()) {
            return null;
        }

        Chamado Chamado = doc.toObject(Chamado.class);
        Chamado.setId(doc.getId());

        return Chamado;
    }

    public List<Chamado> findAll(int limit) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();

        List<Chamado> lista = new ArrayList<>();

        ApiFuture<QuerySnapshot> future = db.collection(COL_NAME)
                .limit(limit)
                .get();

        for (QueryDocumentSnapshot doc : future.get().getDocuments()) {
            Chamado Chamado = doc.toObject(Chamado.class);
            Chamado.setId(doc.getId());

            lista.add(Chamado);
        }

        return lista;
    }
    
    public List<Chamado> findByStatus(String status) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();

        List<Chamado> lista = new ArrayList<>();

        ApiFuture<QuerySnapshot> future = db.collection(COL_NAME)
                .whereEqualTo("status", status)
                .get();

        for (QueryDocumentSnapshot doc : future.get().getDocuments()) {
            Chamado Chamado = doc.toObject(Chamado.class);
            Chamado.setId(doc.getId());

            lista.add(Chamado);
        }

        return lista;
    }

    public void delete(String id) throws ExecutionException, InterruptedException {
        Firestore db = FirestoreClient.getFirestore();

        DocumentReference docRef = db.collection(COL_NAME).document(id);
        DocumentSnapshot snapshot = docRef.get().get();

        if (!snapshot.exists()) {
            throw new RuntimeException("Chamado não encontrado");
        }

        docRef.delete().get();
    }
}

File: FRONTEND/MC/src/app/core/service/chamados/chamados.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable, signal } from '@angular/core';

import { Observable } from 'rxjs';
import { environment } from '../../../../environment/environment';
import { Chamado, ComentarioDTO } from '../../../model/chamados/chamado';

@Injectable({
  providedIn: 'root'
})
export class ChamadosService {

  constructor(private http: HttpClient) { }

  private url = signal(`${environment.url}/chamados`)

  public save(chamado: Chamado): Observable<Chamado> {
    return this.http.post<Chamado>(this.url(), chamado, {});
  }

  public update(chamado: Chamado): Observable<Chamado> {
    return this.http.put<Chamado>(this.url(), chamado, {});
  }

  public adicionarComentario(id: string, comentario: ComentarioDTO): Observable<Chamado> {
    return this.http.post<Chamado>(`${this.url()}/${id}/comentario`, comentario, {});
  }
  
  public findById(id: string): Observable<Chamado> {
    return this.http.get<Chamado>(`${this.url()}/${id}`, {});
  }

  public findAll(limit: number): Observable<Chamado[]> {
    return this.http.get<Chamado[]>(`${this.url()}/all/${limit}`, {});
  }

  public findByStatus(status: String): Observable<Chamado[]> {
    return this.http.get<Chamado[]>(`${this.url()}/status/${status}`, {});
  }




}